src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/RazorBrokeredServiceBase.cs

```diff
@@ -24,10 +27,17 @@ protected RazorBrokeredServiceBase(in ServiceArgs args)
         _serviceBrokerClient = new ServiceBrokerClient(args.ServiceBroker, joinableTaskFactory: null);
         _serverConnection = args.ServerConnection;
         _interceptor = args.Interceptor;
+        _projectQueryServiceFactory = args.ExportProvider.GetExportedValue<ProjectQueryServiceFactory>();
 
         Logger = args.ServiceLoggerFactory.GetOrCreateLogger(GetType());
     }
 
+    protected IProjectQueryService CreateProjectQueryService(RemoteDocumentContext context)
+        => _projectQueryServiceFactory.Create(context.TextDocument.Project.Solution);
+
+    protected IProjectQueryService CreateProjectQueryService(Solution solution)
+        => _projectQueryServiceFactory.Create(solution);
```

Comment:
I'm guessing this is used in other work you are in the middle of?

That might answer some of the thoughts I had about putting the query service on `IDocumentContext`, but I think the same argument could be made for this being an extension method on `Solution`

Reply:
Making an extension method on `Solution` is a good idea, though it lacks discoverability. I'll take another look.

Reply:
Ultimately, I decided to introduce `RemoteSolutionSnapshot` and implement `IProjectQueryService` on that. Then, I added a convenience property on `RemoteDocumentContext` to make it easier for existing services to get an `IProjectQueryService` to pass around.